### PR TITLE
Make QuestionnaireSchema immutable

### PR DESCRIPTION
**Pull request by MebinAbraham**
### What is the context of this PR?
Following on from the initial spike https://github.com/ONSdigital/eq-questionnaire-runner/pull/208, this PR makes the `QuestionnaireSchema` immutable.

__Additional changes__
The existing questionnaire schema mutates the questionnaire JSON on initialisation to add parent ids to the relevant properties. I have refactored this so that the questionnaire JSON is no longer mutated by keeping a map of parent ids. This highlighted an issue in validator in which duplicate ids between blocks and questions are allowed. This has been addressed by: https://github.com/ONSdigital/eq-questionnaire-validator/pull/54
Therefore, some Census schemas will break part way until https://github.com/ONSdigital/eq-questionnaire-schemas/pull/144 has been merged/released and updated in runner.

### How to review 
Ensure:
- app still behaves as expected.
- `QuestionnaireSchema` is in fact immutable.
- changes are suitable and make sense.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

### CODE REVIEW

File: app/forms/field_handlers/field_handler.py [hunk 1]
```diff
@@ -4,8 +4,8 @@
 from wtforms import validators, Field
 
 from app.data_model.answer_store import AnswerStore
-from app.questionnaire.location import Location
 from app.forms.validators import ResponseRequired, format_message_with_title
+from app.questionnaire.location import Location
```
Comment (JamesGardiner) [inline app/forms/field_handlers/field_handler.py:8]: Any chance we could put Location into the Questionnaire __init__.py and change this to:

```suggestion
from app.questionnaire import Location
```

File: app/questionnaire/placeholder_renderer.py [hunk 2]
```diff
@@ -57,29 +55,36 @@ def render_placeholder(self, placeholder_data, list_item_id):
...
-        return placeholder_data["text"].format(**transformed_values)
+        return mutable_placeholder_data["text"].format(**transformed_values)
 
     def render(self, dict_to_render, list_item_id):
         """
         Transform the current schema json to a fully rendered dictionary
         """
-        rendered_data = deepcopy(dict_to_render)
-        pointers = find_pointers_containing(rendered_data, "placeholders")
+        dict_to_render = QuestionnaireSchema.get_mutable_deepcopy(dict_to_render)
```
Comment (JamesGardiner) [inline app/questionnaire/placeholder_renderer.py:83]: `__deepcopy__` should be implemented on QuestionnaireSchema:
```suggestion
        dict_to_render = deepcopy(QuestionnaireSchema(dict_to_render))
```

File: app/questionnaire/questionnaire_schema.py [hunk 3]
```diff
@@ -421,28 +478,11 @@ def _get_error_messages(self):
...
-    nested_objects = OrderedDict()
-
-    for parent_id, child_object in parent_object.items():
-        for child_list_object in child_object.get(list_key, []):
-            # patch the ID of the parent onto the object
-            child_list_object["parent_id"] = parent_id
-            nested_objects[child_list_object["id"]] = child_list_object
-
-    return nested_objects
+def is_hashable(value):
```
Comment (JamesGardiner) [inline app/questionnaire/questionnaire_schema.py:481]: Idiomatic way to check is hashable is to do `isinstance(obj, collections.abc.Hashable)`

File: app/questionnaire/questionnaire_schema.py [hunk 4]
```diff
@@ -421,28 +478,11 @@ def _get_error_messages(self):
...
+    return hasattr(value, "__hash__") and callable(value.__hash__)
 
 
-def _get_values_for_key(block, key, ignore_keys=None):
+def get_values_for_key(block, key, ignore_keys=None):
```
Comment (JamesGardiner) [inline app/questionnaire/questionnaire_schema.py:485]: If we are making this a public method it should probably go into schema utils module

File: app/questionnaire/placeholder_parser.py [hunk 5]
```diff
@@ -33,6 +34,7 @@ def __init__(
         self._placeholder_map = {}
 
     def __call__(self, placeholder_list: Sequence[Mapping]) -> Mapping:
+        placeholder_list = QuestionnaireSchema.get_mutable_deepcopy(placeholder_list)
```
Comment (JamesGardiner) [inline app/questionnaire/placeholder_parser.py:37]:
```suggestion
        placeholder_list = deepcopy(QuestionnaireSchema(placeholder_list))
```

File: app/questionnaire/placeholder_renderer.py [hunk 6]
```diff
@@ -57,29 +55,36 @@ def render_placeholder(self, placeholder_data, list_item_id):
             list_store=self._list_store,
         )
 
-        if "text_plural" in placeholder_data:
-            plural_schema = placeholder_data["text_plural"]
+        mutable_placeholder_data = QuestionnaireSchema.get_mutable_deepcopy(
```
Comment (JamesGardiner) [inline app/questionnaire/placeholder_renderer.py:58]:
```suggestion
        mutable_placeholder_data = deepcopy(QuestionnaireSchema(
```

Comment (JamesGardiner) [reply app/questionnaire/placeholder_renderer.py:58]: Also, is there any need to keep the reference to the placeholder_data separate to the mutable_placeholder_data?

Wouldn't:

```Python
placeholder_data = deepcopy(QuestionnaireSchema(placeholder_data))
```

be more readable?

**Comment by JamesGardiner** [comment]
I've resolved my comments re `deepcopy` - we shouldn't be implementing the `__deepcopy__` method on this object, as we need to return a different object

File: app/forms/field_handlers/field_handler.py (hunk 1 above)
Comment (MebinAbraham) [reply app/forms/field_handlers/field_handler.py]: Done

File: app/questionnaire/placeholder_renderer.py (hunk 6 above)
Comment (MebinAbraham) [reply app/questionnaire/placeholder_renderer.py]: Nope, changed.

File: app/routes/schema.py [hunk 7]
```diff
@@ -12,7 +14,7 @@
 def get_schema_json_from_name(schema_name):
     try:
         schema = load_schema_from_name(schema_name)
-        return jsonify(schema.json)
+        return jsonify(QuestionnaireSchema.get_mutable_deepcopy(schema.json))
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply app/routes/schema.py:17]: Will remove, it was left over from the `Immutables.Map` as jsonify didn't like that.

**Pull request merged**